Wildcard ignores

darkowlzz opened the pull request:
<!--
Work-in-progress PRs are welcome as a way to get early feedback - just prefix
the title with [WIP].
-->

### What does this do / why do we need it?
Adds support for ignoring packages and subpackages using wildcard notation ("/*").

### What should your reviewer look out for in this PR?
Implementation and tests. All the changes are in manifest, not touching gps.

### Do you need help or clarification on anything?
Do we use "/*" or "/..." for this need?

### Which issue(s) does this PR fix?

fixes #691 

<!--

fixes #
fixes #

-->

sdboyer commented:
we definitely want `/*`, as the rules we're applying aren't the same as `/...`.

this approach of doing it out in dep isn't going to work, unfortunately. at most, it helps us with the root project, but it doesn't do anything for dependencies. we've gotta handle this in the solver.

sdboyer commented:
specifically, we'll need to introduce some new props, maybe a slice of strings, on the `rootdata` struct that hold the prefixes to exclude. then, everywhere that we currently check `rd.ig` in the solver...well, we can convert those to a new method, `rootdata.isIgnored(path string)`, which should check both the literal `rd.ig` map and do a prefix-based exclusion check on each of the strings.

actually, no, not a slice. we'll want a trie, so that the prefix search is O(1) in the number of wildcard ignores specified by the root.

i'm musing on whether or not we want the ignores to be case-insensitive, in keeping with #1079. for now, let's assume no - not only is that easier, but i think it's more consistent with shell-globbing, and so will be less confusing for people.

darkowlzz commented:
ah! right! didn't think of the dependencies. I'll try to implement it in the solver 😊

darkowlzz commented:
Another attempt 😅
Removed the first commit of this PR which changed manifest, but kept the ensure integration test which tested recursive ignore in the main project.

sdboyer commented on internal/gps/pkgtree/pkgtree.go:
```diff
@@ -16,8 +16,13 @@ import (
 )
 
+// recursive ignore suffix
+const recIgnoreSuffix = "/*"
```
i don't see a reason why we actually have to restrict to immediately following a slash - i think our only criteria here is that the asterisk has to be the last character.

sdboyer commented on internal/gps/pkgtree/pkgtree.go:
```diff
@@ -445,6 +450,15 @@ func (t PackageTree) ToReachMap(main, tests, backprop bool, ignore map[string]bo
 	}
 
+	// Create a radix tree for ignore prefixes
+	xt := radix.New()
```
this allocs because it returns a pointer, so let's defer creating the trie until the last possible moment - _if_ we see something with the expected suffix.

darkowlzz commented:
* Renamed to "wildcard" and changed "/*" to "*" only.
* Create new radix tree only when wildcard suffix is found in ignore list.
* Updated docs/Gopkg.toml.md with wildcard ignore.

Do we have any existing repo that could be used to test dependency package ignore?

sdboyer commented on internal/gps/pkgtree/pkgtree.go:
```diff
@@ -445,6 +450,20 @@ func (t PackageTree) ToReachMap(main, tests, backprop bool, ignore map[string]bo
+			if xt == nil {
+				xt = radix.New()
+			}
+			i = strings.TrimSuffix(i, wcIgnoreSuffix)
```
This ends up on the `strings.HasSuffix()` call. better to just do the trim operation ourselves.

darkowlzz replied:
You mean for cases like `"github.com/x/someprefix-*-somesuffix"`?
Like using regex match?

sdboyer replied:
nope, i meant that this is the implementation of `string.TrimSuffix()`:

```go
func TrimSuffix(s, suffix string) string {
	if HasSuffix(s, suffix) {
		return s[:len(s)-len(suffix)]
	}
	return s
}
```

and the conditional has already been checked, so it's better to just do the `s[:len(s)-len(suffix)]` operation ourselves rather than calling `strings.TrimSuffix()`, as this method is processing-bound (possibly memory, possibly CPU), and this is within a loop, so easy microoptimizations like this are worth it.

sdboyer commented:
so the PR as-is seems like it covers the essentials. now, i think, we need to do the harder work of thinking through the models and making sure that the edge cases are covered - especially the new ones that are induced by this.

here are a few i can think of:

* i think a global ignore - "*" - should probably just not be allowed.
* ineffectual ignores should be something we warn on, or at the _very_ least do not include in hash inputs. that is, if `github.com/foo*` is ignored, then `github.com/foo/bar*` is ineffectual.
* we should definitely have at least one more hash inputs test.
* is there anywhere that we checks against ignored lists up in dep itself? perhaps on `ensure -add`. if so, they'll need to be updated. and we may need to add some new ones, though i'm not exactly sure what that would look like.

these are just what i can think of right now.

also, this change will probably entail us bumping the solver version. consequently, i may want to wait to merge it until just before we're ready to make the next release (which'll be `v0.4.0`), as otherwise we potentially create contention between team members who're chasing tip, vs. those using a release version.

darkowlzz commented:
- [x]  ignore global ignores
- [x] ignore ineffectual wildcard ignores
- [x] do not include ineffectual wildcard ignores in hash inputs
- [x] hash inputs test to verify exclusion of ineffectual wildcard ignores

> is there anywhere that we checks against ignored lists up in dep itself? perhaps on ensure -add. if so, they'll need to be updated. and we may need to add some new ones, though i'm not exactly sure what that would look like.

@sdboyer  Yes, right now we only use `IgnoredPackages` in `ensure -add`. But since this is passed to `pkgtree.ToReachMap()`, and `ToReachMap()` being aware of how to handle wildcards, there shouldn't be any issues.

I have a [PR](https://github.com/golang/dep/pull/1216/files#diff-cf328f431a592d0b360d509c6c5862c7R607) that uses `IgnoredPackages` but I think that can be avoided by using `ToReachMap()` and get package maps with all the ignoring rules applied.

darkowlzz commented:
I went through both gps and dep, and found that most of the places where we care about ignoring use `PackageTree.ToReachMap()` with the ignored list passed to them to get a map of packages. So, no change would be required in those places.

[`SolveParameters.toRootdata()`](https://github.com/golang/dep/blob/master/internal/gps/solver.go#L212) checks if a package is present in `required` as well as `ignored`. That needs to be changed to check the wildcard ignores too. Will add that in the next commit.

sdboyer commented on internal/gps/pkgtree/pkgtree.go:
```diff
@@ -16,8 +16,13 @@ import (
 	"strings"
 	"unicode"
+
+	radix "github.com/armon/go-radix"
```
the package name is already `radix`, so this is unnecessary.

darkowlzz commented:
Excited for this 😃

ibrasho commented on internal/gps/pkgtree/pkgtree.go:
```diff
@@ -928,3 +944,42 @@ func uniq(a []string) []string {
+// CreateIgnorePrefixTree takes a set of strings to be ignored and returns a
+// trie consisting of strings prefixed with wildcard ignore suffix (*).
+func CreateIgnorePrefixTree(ig map[string]bool) *radix.Tree {
+	var xt *radix.Tree
```
Why not initialize a new tree here instead of checking for nil later on?

darkowlzz replied:
We had that earlier but then we decided to allocate memory only when we actually get a wildcard suffix. Refer https://github.com/golang/dep/pull/1156#discussion_r138512356

sdboyer commented:
woohoo!

i think we should roll a release after this. speaking of, this is missing a CHANGELOG update :)

sdboyer commented:
fix CHANGELOG conflict for this, and we're good to go on it

sdboyer commented:
nvm, i just fixed it directly :)
